### DISCUSSION TITLE
IM7 CLUT and handling alpha

### DISCUSSION THREAD

**Discussion by MMz10000** [ImageMagick/ImageMagick#2079] [Help] [answered]
Good afternoon,

I'm having trouble with the "CLUT" option.
My portable IM version is ImageMagick 7.0.10-15 Q16 x64 2020-05-30 running on windows 10 x64 (using powershell)

I am trying to follow the example provided by Anthony ([CLUT and transparency handling](https://www.imagemagick.org/Usage/color_mods/#clut_alpha)), but those are tailored for IM6, and I came across many post describing how IM7 is handling alpha instead of the opacity. I first tried to understand what was explained, then tried to brute-force a solution, but my very limited experience couldn't solve the situation.

So today, I am writing this post to first know how to "translate" Anthony's example from IM6 to IM7, so I can try to learn what are the differences and what do they mean (or if you can explain them, it would be fantastic).

But in the end, my objective is to generate a multi gradient with linear smooth transition between colors, and I need it to handle transparency for each color I'm passing (for example "#FFFFFFFF" > "#FFFFFF00" > "#FFFFFFFF", where the middle of the gradient would be perfectly transparent.
As I don't understand how to attach files, I can't show you what I am trying to achieve...

Could someone help me please?

**Comment by fmw42** [comment]
To post an image, just drag and drop into the box. We need to see an example to understand your request.

**Comment by fmw42** [comment]
I do not believe there is a close analogy in IM 7. This is at least one way to replicate Anthony's result. You have to copy the image to make an alpha channel. Then save it in mpr: format and delete the copy. Then after the clut processing add the alpha channel back.

`magick blurred_shape.jpg ( +clone -negate -threshold 40% +write mpr:alpha +delete ) gradient_border.png -interpolate integer -clut mpr:alpha -alpha off -compose copy_opacity -composite clut_shape7.png`

ADDITION:

Actually this is not really a good approach. It gets the alpha from the input and not the clut. I will have to review Anthony's command further.

**Comment by MMz10000** [comment]
Good afternoon @fmw42 

All right, I didn't check my web browser configuration, it should be fine now...

Base image: `magick -size 256x128 -define gradient:extent=ellipse radial-gradient:white-black`
![Base](https://user-images.githubusercontent.com/47039265/83334372-fdcf2e00-a29d-11ea-8ba4-ae362e542a3e.png)

Output obtained with: `magick base.png clutX.png -clut output.png`
CLUT A (last color being #FF00FFFF) and output A:
![Clut A](https://user-images.githubusercontent.com/47039265/83334410-45ee5080-a29e-11ea-8bfd-a3b8bd289c7e.png) => ![Output A](https://user-images.githubusercontent.com/47039265/83334419-53a3d600-a29e-11ea-8063-d278d6ea601f.png)

CLUT B (last color being #FF00FF00) and output B:
![Clut B](https://user-images.githubusercontent.com/47039265/83334454-96fe4480-a29e-11ea-8bc2-c830ec7d57a6.png) => ![Output B](https://user-images.githubusercontent.com/47039265/83334458-9e255280-a29e-11ea-9f19-0528973f8681.png)

You will notice strong aliasing in Output B, but I hope it will go away when the alpha problem is fixed...

Also thanks for all the help you provide, I came across your answers so often!

**Reply by fmw42** [reply]
What color or region do you want to be transparent here?

**Comment by MMz10000** [comment]
To answer your suggestion, would it be a good solution to process my image without alpha at all, just on the RGB channels, and then generate a gray-scaled mask using the CLUT in the same fashion, but only on the alpha channel. To finally apply that mask on my first output?

Regards

**Reply by fmw42** [reply]
Yes, you could create a mask for where you want the result to be transparent. Then process the gradient with no transparency. Then apply your mask at the end to the result by putting it in the alpha channel of the output. That is essentially what I did in my answer above.

**Reply by MMz10000** [reply]
I will try this and once I'm done, post the result here. If you think the code is nice enough, then I will close this thread if you don't mind checking.
Thanks again!

**Comment by MMz10000** [comment]
I actually have trouble to generate a good CLUT :(
Here are the color I try to build my CLUT with: "#ff0000ff", "#00ff00ff", "#0000ffff", "#ffffffff", "#ff00ff00"

I'm build my CLUT by `-append` the following outputs in my script: `-size 1x$($currentOffset) gradient:$($Colors[$i+1])-$($Colors[$i])`

Here is the output:
![clut](https://user-images.githubusercontent.com/47039265/83335167-b26b4e80-a2a2-11ea-9e57-e4c6808974e1.png)

Opening it on GIMP reveals that the purple is completely gone. Curiously, setting the final color as "#ff00ff00" give the following CLUT:
![clut](https://user-images.githubusercontent.com/47039265/83335223-23126b00-a2a3-11ea-93df-a871e7351f9c.png)
Which colors ONLY the final pixel as purple:
![image](https://user-images.githubusercontent.com/47039265/83335234-37eefe80-a2a3-11ea-8caf-30f41be48363.png)

I would have expected in both case that the white would have linearly turned to purple anyway, alpha channel or not, in both outputs... Do you have a clue?

Considering the CLUT is wrong at the moment, I was wondering if the following line was still correct: `base.png `( clut.png -alpha remove `) -clut RGB.png` ?
Here is the current ouput with the CLUT containing "#FF00FF0**1**" (see screenshot above)
![RGB](https://user-images.githubusercontent.com/47039265/83335310-cfece800-a2a3-11ea-9c7d-7ffc9084e30b.png)

**Reply by fmw42** [reply]
Please show the full code. I do not understand some of your items. Why are you doing -size 1x$($currentOffset) rather than -size 1x$currentOffset?  What happens if you just build manually rather than in a script with the few values you have?  What image format are you using to save the clut image?

**Reply by MMz10000** [reply]
Here is my function (can't get the markdown to work properly):

function CustomGradientRadial ( [string]$ExportName, [string[]]$Colors, [decimal[]]$OffsetsUser) {
    [decimal[]]$Offsets = @(0) + $OffsetsUser
    [int]$previousOffset = 0
    [int]$currentOffset = 0
    [string]$clutCommand = ""

    for ( $i=$Offsets.Length-1; $i -ge 0; $i--) {
        $currentOffset = (100-[math]::Truncate($Offsets[$i]*100)) - $previousOffset
        $clutCommand += "``( -size 1x$($currentOffset) gradient:$($Colors[$i+1])-$($Colors[$i]) ``) "
        $previousOffset = 100-[math]::Truncate($Offsets[$i]*100)
    }

    im -size 256x128 -define gradient:extent=ellipse radial-gradient:white-black $dir_utils\base.png
    Invoke-Expression "im ``( $clutCommand -append ``) $dir_utils\clut.png"

# Testing from here
im $dir_utils\base.png `( $dir_utils\clut.png -alpha remove `) -clut $dir_utils\RGB.png

}

The code stops here...
And I'm calling it like this:
`CustomGradientRadial "Output" @( "#ff0000ff", "#00ff00ff", "#0000ffff", "#ffffffff", "#ff00ff00") @( 0.25, 0.5, 0.75)`

I'm closer to be a beginner concerning scripting and programming... I'm not a professional, nor an amateur either :D For whatever reasons, `-size 1x$currentOffset` doesn't work in powershell, but `-size 1x$($currentOffset)` does. I can't explain why, but as much as I really don't like powershell, I think it's still better than batch...

I will try to forge a clut image manually and post the results, but obviously, I need this function to work. And so far, I'm working with miff, but using png for my final outputs and test pictures. So the function uses png exclusively so far.

**Reply by fmw42** [reply]
Sorry, I though your code was Unix bash shell scripting. I am not a Windows user, so cannot help with your code. But try it manually. You do not have that many values. Post that code if you do not get expected results. But currently I do not think you can use transparency in cluts in IM 7 as Anthony did it in IM 6.

**Reply by MMz10000** [reply]
Here is the CLUT I try scripting. The following has been made under GIMP.
With alpha channel on:
![image](https://user-images.githubusercontent.com/47039265/83336045-9074ca80-a2a8-11ea-8307-b6bafb57983d.png)

And the very same picture but with the alpha channel off:
![image](https://user-images.githubusercontent.com/47039265/83336064-9f5b7d00-a2a8-11ea-8840-0b60f905ffa3.png)

Using this CLUT, here is my test output:
![Output](https://user-images.githubusercontent.com/47039265/83336183-45a78280-a2a9-11ea-8ca2-436a82f0d9d2.png)

If I succeed to get that by scripting only, then I would generate the mask in the same fashion separately and apply it on this current output to get the desired gradient, handling transparency.

**Comment by fmw42** [comment]
I will look further into an IM 7 equivalent. 

For now, I think you need to create everything with no transparency and then mask your result afterwards to add transparency.

**Reply by MMz10000** [reply]
This is what I try to do. I tried explaining that there: https://github.com/ImageMagick/ImageMagick/discussions/2079#discussioncomment-20418

**Reply by MMz10000** [reply]
So aftre forging my CLUT in GIMP, I did some ugly script and the result is not far off my expectations.

Here is the code:
magick base.png `( clut.png -alpha remove `) -clut RGB.png
magick base.png `( clut.png -alpha extract `) -clut alpha.png
magick composite -compose CopyOpacity alpha.png RGB.png result.png

And here is the output:
![result](https://user-images.githubusercontent.com/47039265/83336541-442b8980-a2ac-11ea-9aa4-a6db99a45ed2.png)
Zoomed:
![image](https://user-images.githubusercontent.com/47039265/83336606-c74cdf80-a2ac-11ea-8ae1-19b26e9777b7.png)
It sorts of fade to transparent purple :D

But opening result.png on GIMP revealed something I wasn't expecting. Here is a screenshot with the alpha channel disabled:
![image](https://user-images.githubusercontent.com/47039265/83336563-6f15dd80-a2ac-11ea-8367-d501ca1ee3f5.png)

It fades to purple, but then to black instead of keeping the purple information in the transparent pixels.

I was expecting seeing the following (with the alpha channel disabled of course):
![image](https://user-images.githubusercontent.com/47039265/83336597-bac88700-a2ac-11ea-8a28-ffd08789870f.png)

Once I get this to work, I will use a better syntax, inspiring myself from the translation you provided there later (https://github.com/ImageMagick/ImageMagick/discussions/2079#discussioncomment-20442).

**Reply by fmw42** [reply]
So what is wrong or not as expected?

**Reply by MMz10000** [reply]
You have completely solved the problem there https://github.com/ImageMagick/ImageMagick/discussions/2079#discussioncomment-20445

But to explain better, you can see that disabling the alpha channel in GIMP revealed that the purple information was discarded on the transparent pixels.

**Reply by fmw42** [reply]
I do not know how GIMP handles transparency nor what file format you use to save your result. If you do all the processing in ImageMagick, then do 

magick image.png -alpha off result.png

Does that show what you expect? Also what color was in the clut under your transparency.

**Comment by fmw42** [comment]
Here is how to replicate Anthony's IM 6 code in IM 7. You have to process the image with clut without transparency and then process the image with the clut's transparency only. Then put the latter into the alpha channel of the former.

Image:

![blurred_shape](https://user-images.githubusercontent.com/13458218/83336416-13c1fc00-a268-11ea-9453-5bdb9e0d5398.jpg)


Clut:

![gradient_border](https://user-images.githubusercontent.com/13458218/83336419-1886b000-a268-11ea-8978-253daa62c24a.png)

Windows code:

```
magick ^
( blurred_shape.jpg -write mpr:img +delete ) ^
( gradient_border.png -write mpr:clut +delete ) ^
( mpr:img mpr:clut -alpha off -interpolate integer -clut ) ^
( mpr:img -alpha off ( mpr:clut -alpha extract ) -interpolate integer -clut ) ^
-alpha off -compose copy_opacity -composite ^
clut_shape7.png

```

![clut_shape7](https://user-images.githubusercontent.com/13458218/83336440-4bc93f00-a268-11ea-904f-50f94d4fee6f.png)

**Comment by fmw42** [comment, marked as answer]
Here is a simpler way. Just put a copy of the image into its own alpha channel so that the alpha channel of the clut can process the alpha channel of the image. Windows syntax.

`magick ( blurred_shape.jpg -alpha copy ) gradient_border.png -channel RGBA -interpolate integer -clut clut_shape7.png`

**Reply by MMz10000** [reply]
Incredible!

Here is what I did in my script:
`magick ( grayscale.png -alpha copy ) gradient_clut.png -channel RGBA -clut output.png`

And here are the images used:
grayscale.png
![grayscale](https://user-images.githubusercontent.com/47039265/83338180-9cb55380-a2b9-11ea-8879-65b93ec0d594.png)
and gradient_clut.png
![gradient_clut](https://user-images.githubusercontent.com/47039265/83338186-b35baa80-a2b9-11ea-981b-7792478d4444.png)

GIMP screenshots.
Alpha enabled:
![image](https://user-images.githubusercontent.com/47039265/83336747-e730d300-a2ad-11ea-9da7-274e0d43074d.png)

Alpha disabled:
![image](https://user-images.githubusercontent.com/47039265/83336750-edbf4a80-a2ad-11ea-9971-c3fb40159e6c.png)

Well done! This is the final expected result!

Now, my only problem is to generate the CLUT properly.

**Reply by MMz10000** [reply]
I will start another thread as it is not related to this topic anymore.
Before marking it as an answer, do you think you can explain the magick behind this line, step by step, so everyone can understand, please?

**Reply by fmw42** [reply]
In Anthony's example, the input image had no alpha channel. But in IM 6, the alpha channel in clut could work on the opaque image. But in IM 7, there has to be an alpha channel in the image for the alpha channel of the clut to process it. So one has to copy the image into its alpha channel to get the same result in IM 7.

**Reply by MMz10000** [reply]
Awesome! Thanks for the full explanation.

I'm very grateful for the time you have spent helping me.
